src/Rules/Operators/InvalidBinaryOperationRule.php

```diff
@@ -105,6 +106,25 @@ public function processNode(Node $node, Scope $scope): array
 				return [];
 			}
 
+			if ($node instanceof Node\Expr\BinaryOp\Mod) {
+				$leftNumberType = $leftType->toNumber();
+				$rightNumberType = $rightType->toNumber();
+				if (!$leftNumberType instanceof ErrorType && !$rightNumberType instanceof ErrorType) {
```

Comment:
So just change the rule please. The condition should be written differently.

1) Please verify that all the `X` from table here https://github.com/phpstan/phpstan-src/pull/2466 are reported with `'Binary operation "%s" between %s and %s results in an error.'`.
2) Please write tests that all the `D` are reported with `'Deprecated in PHP 8.1: Implicit conversion from %s to %s loses precision.'`.
3) Please note that types can for example be `int|float` or `mixed`. You should be able to take advantage of RuleLevelHelper so that something that's always wrong like `int % float` is always reported, something that might be wrong like `int % float|int` is reported on level 7 and `int % mixed` is reported on level 9.

Please see how `RuleLevelHelper::findTypeToCheck` is used in many existing rules like InvalidCastRule. Basically there's a callback that checks the type and transforms the type based on the rule level. Then there's `$type instanceof ErrorType` with early return and then there's the callback called again on the transformed type.